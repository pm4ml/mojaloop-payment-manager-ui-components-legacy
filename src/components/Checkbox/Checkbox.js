import './Checkbox.scss';

import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

import * as utils from '../../utils/common';

class Checkbox extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      checked: this.props.checked,
    };
    this.onClick = this.onClick.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onFocus = this.onFocus.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.testKey = this.testKey.bind(this);
  }
  componentDidUpdate(prevProps) {
    if (this.props.checked !== prevProps.checked) {
      // eslint-disable-next-line
      this.setState({ checked: this.props.checked });
    }
  }

  onChange(e) {
    e.preventDefault();
    e.stopPropagation();
    if (this.props.disabled) {
      return;
    }

    const checked = !this.state.checked;
    this.setState({ checked });
    if (this.props.onChange) {
      this.props.onChange(checked);
    }
  }
  onBlur(e) {
    if (this.props.onBlur) {
      this.props.onBlur(e);
    }
  }
  onClick(e) {
    e.preventDefault();
    e.stopPropagation();
    if (this.props.onClick) {
      this.props.onClick(e);
    }
  }
  onFocus(e) {
    if (this.props.onFocus) {
      this.props.onFocus(e);
    }
  }
  testKey(e) {
    if (e.keyCode === 9) {
      e.stopPropagation();
      e.preventDefault();
      utils.focusNextFocusableElement(this.input, !e.shiftKey);
      return;
    }
    if (e.keyCode === 13) {
      e.stopPropagation();
      e.preventDefault();
      this.onChange(e);
    }
  }
  render() {
    const { checked } = this.state;
    const { style, className, id, label, disabled, round, semi } = this.props;
    const checkboxClassName = utils.composeClassNames(['mb-input', 'input-checkbox', className]);
    const labelClassName = utils.composeClassNames([
      round && 'input-checkbox__label--round',
      !label && 'input-checkbox__label--no-margin',
    ]);
    return (
      <div className={checkboxClassName} style={style}>
        <input
          ref={input => {
            this.input = input;
          }}
          type="checkbox"
          id={id}
          className={`input-checkbox__input ${semi ? 'semi-checked' : ''}`}
          onClick={this.onClick}
          onBlur={this.onBlur}
          onFocus={this.onFocus}
          onKeyDown={this.testKey}
          onChange={e => {
            e.preventDefault();
            e.stopPropagation();
          }}
          checked={checked && semi !== true}
          disabled={disabled}
        />
        {/* eslint-disable-next-line */}
        <label htmlFor={id} className={labelClassName} onClick={this.onChange} role="button">
          {label}
        </label>
      </div>
    );
  }
}
Checkbox.propTypes = {
  style: PropTypes.shape(),
  className: PropTypes.string,
  semi: PropTypes.bool,
  id: PropTypes.string,
  label: PropTypes.string,
  round: PropTypes.bool,
  onBlur: PropTypes.func,
  onClick: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
};
Checkbox.defaultProps = {
  style: undefined,
  className: undefined,
  semi: false,
  id: undefined,
  label: undefined,
  round: false,
  onBlur: undefined,
  onClick: undefined,
  onChange: undefined,
  onFocus: undefined,
  checked: false,
  disabled: false,
};
export default Checkbox;
