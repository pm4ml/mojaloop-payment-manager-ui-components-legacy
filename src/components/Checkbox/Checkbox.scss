@import '../../assets/styles/vars/colors.scss';
@import '../../assets/styles/vars/shadow.scss';

.input-checkbox {
  display: flex;
  align-items: center;
}
.input-checkbox__input {
  width: 0px;
  height: 0px;
  appearance: none;
  opacity: 0;
  box-shadow: 0 1px 2px rgba(0, 0, 0, 0.05), inset 0px -15px 10px -12px rgba(0, 0, 0, 0.05);
  position: relative;
  margin: 0px;
  padding: 0px;

  &:focus {
    + label {
      &:before {
        box-shadow: $shadow-size rgba(0, 175, 255, 0.5);
      }
    }
  }

  + label {
    max-width: 100%;
    display: flex;
    position: relative;
    cursor: pointer;
    align-items: center;
    font-size: 14px;
    user-select: none;
    height: 22px;

    &:hover {
      &:before {
        box-shadow: $shadow-size rgba(0, 0, 0, 0.13);
      }
    }

    &.input-checkbox__label--round {
      &:before {
        border-radius: 12px;
      }
    }
    &:before {
      box-sizing: initial;
      flex: 0 0 20px;
      display: inline-block;
      margin-right: 5px;
      background: linear-gradient(#fafafa, #f4f4f4);
      border: 1px solid #cacbcc;
      content: '';
      width: 20px;
      height: 20px;
      box-shadow: 0px 0px 0px 0px rgba(0, 0, 0, 0.13);
      transition: all 0.2s ease-out;
      flex-basis: 22px;
      width: 22px;
      border: none;
      height: 20px;
      border-bottom: 2px solid rgba(0, 0, 0, 0.2);
    }
    &.input-checkbox__label--no-margin {
      &:before {
        margin-right: 0px;
      }
    }
  }

  &:disabled {
    + label {
      color: #999;
      cursor: default;
      &:before {
        background: #eee;
        box-shadow: none;
      }
    }
    &:checked {
      + label {
        &:before {
          border-color: #60d2fd;
          background: #60d2fd;
        }
      }
    }
  }

  &:checked {
    + label {
      &:before {
        border-color: #00a2df;
        background: linear-gradient(#10b2ef, #00a2df);
      }
      &:after {
        content: '';
        box-sizing: content-box;
        box-sizing: initial;
        position: absolute;
        left: 6px;
        top: 10px;
        background: #fff;
        width: 2px;
        height: 2px;
        box-shadow: 2px 0 0 #fff, 4px 0 0 #fff, 4px -2px 0 #fff, 4px -4px 0 #fff, 4px -8px 0 #fff,
          4px -6px 0 #fff;
        animation: animateCheckboxSelectedContent 0.2s ease-out 1 forwards;
        transform: rotate(45deg);
      }
    }
  }
  &.semi-checked {
    + label {
      &:before {
        border-color: #ddd;
        background: #ddd;
      }

      &:after {
        content: '';
        box-sizing: content-box;
        box-sizing: initial;
        position: absolute;
        left: 6px;
        top: 10px;
        background: #999;
        width: 10px;
        height: 2px;
      }
    }
  }
}

@keyframes animateCheckboxSelectedContent {
  0% {
    margin-top: 2px;
    opacity: 0;
  }
  100% {
    margin-top: 0px;
    opacity: 1;
  }
}
