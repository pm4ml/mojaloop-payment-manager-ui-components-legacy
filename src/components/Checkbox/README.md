# Checkbox

Renders a checkbox.

### Proptypes

|    Name   |   Type   |                 Description                  | Optional | Values |   Default   |
|-----------|----------|----------------------------------------------|----------|--------|-------------|
| classname | `string` | defines the HTML classname                   | x        |        | `undefined` |
| id        | `string` | defines the HTML id                          | x        |        | `undefined` |
| style     | `object` | defines the HTML style                       | x        |        | `undefined` |
| label     | `string` | defines the text content                     |          |        | `undefined` |
| disabled  | `bool`   | defines if the button should be disabled     | x        |        | `false`     |
| checked   | `bool`   | defined whether the checkbox is checked      |          |        | `false`     |
| semi      | `bool`   | defines whether the checkbox is half-checked | x        |        | `false`     |
| onClick   | `func`   | defines the input _onclick_ function         |          |        | `undefined` |
| onChange  | `func`   | defines the input _onchange_ function        |          |        | `undefined` |
| onBlur    | `func`   | defines the input _onblur_ function          |          |        | `undefined` |
| onFocus   | `func`   | defines the input _onfocus_ function         |          |        | `undefined` |
