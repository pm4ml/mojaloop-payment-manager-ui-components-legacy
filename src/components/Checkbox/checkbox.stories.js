/* eslint no-alert: "off" */
/* eslint no-console: "off" */
/* eslint no-console: "off" */
import React from 'react';
import Checkbox from './Checkbox';
import Column from '../Column';

export default {
  title: 'Checkbox',
  component: Checkbox,
};

export const Default = () => (
  <Column align="space-between">
    <div className="m5">
      <Checkbox label="Default Checkbox" checked={false} />
    </div>
    <div className="m5">
      <Checkbox label="Default Checkbox" checked />
    </div>
  </Column>
);

export const SemiChecked = () => (
  <Column align="space-between">
    <div className="m5">
      <Checkbox label="Semi Checked Checkbox" checked={false} semi />
    </div>
    <div className="m5">
      <Checkbox label="Semi Checked Checkbox" checked semi />
    </div>
  </Column>
);

export const Round = () => (
  <Column align="space-between">
    <div className="m5">
      <Checkbox label="Round Checkbox" checked={false} round />
    </div>
    <div className="m5">
      <Checkbox label="Round Checkbox" checked round />
    </div>
  </Column>
);

export const Disabled = () => (
  <Column align="space-between">
    <div className="m5">
      <Checkbox label="Disabled Checkbox" checked={false} disabled />
    </div>
    <div className="m5">
      <Checkbox label="Disabled Checkbox" checked disabled />
    </div>
  </Column>
);

export const OnChange = () => (
  <Column align="space-between">
    <div className="m5">
      <Checkbox label="onChange event Checkbox" checked={false} onChange={console.log} />
    </div>
    <div className="m5">
      <Checkbox label="onChange event Checkbox" checked onChange={console.log} />
    </div>
  </Column>
);

export const OnFocus = () => (
  <Column align="space-between">
    <div className="m5">
      <Checkbox label="onFocus event Checkbox" checked={false} onFocus={console.log} />
    </div>
    <div className="m5">
      <Checkbox label="onFocus event Checkbox" checked onFocus={console.log} />
    </div>
  </Column>
);

export const OnBlur = () => (
  <Column align="space-between">
    <div className="m5">
      <Checkbox label="onBlur event Checkbox" checked={false} onBlur={console.log} />
    </div>
    <div className="m5">
      <Checkbox label="onBlur event Checkbox" checked onBlur={console.log} />
    </div>
  </Column>
);
