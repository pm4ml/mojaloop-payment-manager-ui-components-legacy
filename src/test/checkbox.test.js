import { mount, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import React from 'react';

import Checkbox from '../components/Checkbox';

it('renders the checkbox', () => {
  const wrapper = shallow(<Checkbox />);
  expect(wrapper.find('div.input-checkbox')).toHaveLength(1);
});

it('renders the checkbox value', () => {
  const wrapper = shallow(<Checkbox checked />);
  expect(wrapper.find('input[type="checkbox"]').prop('checked')).toBe(true);
});

it('renders the label', () => {
  const wrapper = shallow(<Checkbox label="test-checkbox" />);
  expect(wrapper.find('label')).toHaveLength(1);
  expect(wrapper.find('label').text()).toBe('test-checkbox');
});

it('applies the prop className', () => {
  const wrapper = shallow(<Checkbox className="test" />);
  expect(wrapper.find('.test')).toHaveLength(1);
});

it('renders the prop id', () => {
  const wrapper = shallow(<Checkbox id="testCheckboxId" />);
  expect(wrapper.find('#testCheckboxId')).toHaveLength(1);
});

it('renders the disabled state', () => {
  const wrapper = shallow(<Checkbox disabled />);
  expect(wrapper.find('input')).toHaveLength(1);
  expect(wrapper.find('input').prop('disabled')).toBe(true);
});

it('renders the round state', () => {
  const wrapper = shallow(<Checkbox round />);
  expect(wrapper.find('label').hasClass('input-checkbox__label--round')).toBeTruthy();
});

// Events

it('triggers onFocus when focused', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(<Checkbox onFocus={mockEvent} />);
  expect(mockEvent).not.toHaveBeenCalled();
  wrapper.find('input').simulate('focus');
  expect(mockEvent).toHaveBeenCalled();
});

it('triggers onBlur when blurred', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(<Checkbox onBlur={mockEvent} />);
  expect(mockEvent).not.toHaveBeenCalled();
  wrapper.find('input').simulate('blur');
  expect(mockEvent).toHaveBeenCalled();
});

it('triggers onChange when changed', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(<Checkbox onChange={mockEvent} />);
  expect(mockEvent).not.toHaveBeenCalled();
  wrapper.find('label').simulate('click');
  expect(mockEvent).toHaveBeenCalled();
});

it('triggers onChange when pressing Enter key', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(<Checkbox onChange={mockEvent} />);
  expect(mockEvent).not.toHaveBeenCalled();
  wrapper.find('input').simulate('keyDown', { key: 'Enter', keyCode: 13, which: 13 });
  expect(mockEvent).toHaveBeenCalled();
  expect(wrapper.find('input[type="checkbox"]').prop('checked')).toBe(true);
});

it('triggers onBlur when blurring', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(<Checkbox onBlur={mockEvent} />);
  expect(mockEvent).not.toHaveBeenCalled();
  wrapper.find('input').simulate('blur', { key: 'Tab', keyCode: 9, which: 9 });
  expect(mockEvent).toHaveBeenCalled();
});

// Snapshot

it('renders the checkbox correctly when multiple props are set', () => {
  const wrapper = shallow(<Checkbox checked id="test-id" />);
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});
