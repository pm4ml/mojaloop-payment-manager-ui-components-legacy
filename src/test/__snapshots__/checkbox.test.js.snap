// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the checkbox correctly when multiple props are set 1`] = `
<div
  className="mb-input input-checkbox"
>
  <input
    checked={true}
    className="input-checkbox__input "
    disabled={false}
    id="test-id"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    type="checkbox"
  />
  <label
    className="input-checkbox__label--no-margin"
    htmlFor="test-id"
    onClick={[Function]}
    role="button"
  />
</div>
`;
